### DISCUSSION TITLE
how to set up hydrogen to play the entire sample even if i don't press the touche anymore

### DISCUSSION THREAD

**Discussion by mathurien** [hydrogen-music/hydrogen#1834] [Hydrogen on Linux]
hi, when i stop pressing a touch of my midi keybiard (akai mpk mini2) the note (for example a open HH) stop playing, even if the sample is not finished playing, how can i change that so it would play the entire note even if i stop pressing the touch ?

ps: sorry for my english i'm french

**Comment by cme** [comment]
Hi :) You can do this by selecting `Preferences > MIDI System > Ignore note-off`.

Now that I think about it, this would probably be handy to be able to set on a per-instrument basis...

**Reply by elpescado** [reply]
This _should_ be controlled by ASDR envelope on per-instrument basis. Setting "Release" to an arbitrarily high value should result in sample being played after note-off message.

**Reply by theGreatWhiteShark** [reply]
> Now that I think about it, this would probably be handy to be able to set on a per-instrument basis...

Within the Preferences or within the InstrumentEditor? The former - which is currently used to hold this property - affects all songs. Also those created in the past.

> This should be controlled by ASDR envelope on per-instrument basis. Setting "Release" to an arbitrarily high value should result in sample being played after note-off message.

All drumkits available online do have a `release` larger than 0. This will also be present in almost all .h2songs created in the recent years.

Apart from this the UX wouldn't be clear if one has to set the (seemingly unrelated) Release to 0 in order to ignore note-off events.

**Reply by cme** [reply]
Yeah, perhaps "note off" is a bit counterintuitive, and "ADSR" has always felt a bit weird since they're sample-based and so inherently have their own envelopes which ADSR can only further constrain. There's no possibility of sustain beyond what the sample naturally contains.

I agree that setting an arbitrarily long release on relevant instruments *should* do the right thing at least, but for some reason, the release  is capped at about 10k frames. I guess the user interface somewhat dictates this: if it's a knob, it has to have some idea of an upper limit. But perhaps the implementation should treat that upper limit as infinite?

The user point of view is probably a lot simpler, like "this should keep going", especially since we're generally modelling percussion. In fact, all the instruments in the standard kit feel like they should have the 'infinite release' property.

Is there *actually* a use-case for treating note-off's the way they're treated now? Should "ignore note-off" just be default?

**Reply by theGreatWhiteShark** [reply]
> Is there actually a use-case for treating note-off's the way they're treated now? Should "ignore note-off" just be default?

cymbal chokes. The problem in here was that the MIDI device was not an edrum but a MIDI keyboard. (But I own the same and scratched my head in the beginning too)

> "ADSR" has always felt a bit weird since they're sample-based and so inherently have their own envelopes which ADSR can only further constrain

It's pure speculation and I haven't checked, but I guess Hydrogen was not as sample-based way, way back as it is now. There is the `Synth` class, which was more than dead code in the beginning of the git history. Also, its the first row in the InstrumentEditor.

At least for all the kits we provide on SourceForge IIRC there is just a _single_ one which sets some of the ADSR values. All others just keep them on their default settings. Maybe, instead of giving them new meaning they could die or at least be hidden at some point.

Per default all frames of each sample is played without being altered (by the ADSR) and the user has the option to shorten the length of a note/sample using right-click + drag. This mimics NOTEON + NOTEOFF quite well. Viewed from this angle supporting the latter integrates quite nicely.

**Comment by theGreatWhiteShark** [comment]
@mathurien the reason you have to check this option is that your keyboard sends two distinct MIDI events: a `NOTE ON` triggering playback when pressing down a key and a `NOTE OFF` stopping playback when releasing it again.
